File: lib-tcnet/src/tcnet.cpp
/**
 * @file tcnet.cpp
 *
 */
/* Copyright (C) 2019-2020 by Arjan van Vught mailto:info@orangepi-dmx.nl
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:

 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.

 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include <stdint.h>
#include <string.h>
#include <cassert>

#include "tcnet.h"
#include "tcnettimecode.h"

#include "tcnetpackets.h"

#include "hardware.h"
#include "network.h"

#include "debug.h"

static constexpr char NODE_NAME_DEFAULT[] = "AvV-OPi";

TCNet *TCNet::s_pThis = nullptr;

TCNet::TCNet(TTCNetNodeType tNodeType) {
	assert(s_pThis == nullptr);
	s_pThis = this;

	memset(&m_tOptIn, 0, sizeof(struct TTCNetPacketOptIn));

	// Fill the static fields for Opt-IN
	m_tOptIn.ManagementHeader.ProtocolVersionMajor = 3;
	m_tOptIn.ManagementHeader.ProtocolVersionMinor = 3;
	memcpy(m_tOptIn.ManagementHeader.Header, "TCN", 3);
	m_tOptIn.ManagementHeader.MessageType = TCNET_MESSAGE_TYPE_OPTIN;
	m_tOptIn.ManagementHeader.SEQ = 0;
	m_tOptIn.ManagementHeader.NodeType = tNodeType;
	m_tOptIn.ManagementHeader.NodeOptions = 0;
	m_tOptIn.NodeCount = 1;
	m_tOptIn.NodeListenerPort = TCNETUnicast::PORT;
	memcpy(&m_tOptIn.VendorName, "orangepi-dmx.org", TCNET_VENDOR_NAME_LENGTH);
	memcpy(&m_tOptIn.DeviceName, "LTC SMPTE Node  ", TCNET_DEVICE_NAME_LENGTH);
	m_tOptIn.DeviceMajorVersion = static_cast<uint8_t>(_TIME_STAMP_YEAR_ - 2000);
	m_tOptIn.DeviceMinorVersion = _TIME_STAMP_MONTH_;
	m_tOptIn.DeviceBugVersion = _TIME_STAMP_DAY_;

	SetNodeName(NODE_NAME_DEFAULT);
	SetLayer(TCNetLayer::LAYER_M);
	SetTimeCodeType(TCNET_TIMECODE_TYPE_SMPTE_30FPS);
}

TCNet::~TCNet() {
	Stop();
}

void TCNet::Start() {
	m_tNode.nIPAddressLocal = Network::Get()->GetIp();
	m_tNode.nIPAddressBroadcast = m_tNode.nIPAddressLocal | ~(Network::Get()->GetNetmask());

	m_aHandles[0] = Network::Get()->Begin(TCNetBroadcast::PORT_0);
	assert(m_aHandles[0] >= 0);

	m_aHandles[1] = Network::Get()->Begin(TCNetBroadcast::PORT_1);
	assert(m_aHandles[1] >= 0);

#if defined(USE_PORT_60002)
	m_aHandles[2] = Network::Get()->Begin(TCNetBroadcast::PORT_2);
	assert(m_aHandles[2] >= 0);
#endif

#if defined(USE_PORT_UNICAST)
	m_aHandles[3] = Network::Get()->Begin(TCNETUnicast::PORT);
	assert(m_aHandles[3] >= 0);
#endif
}

void TCNet::Stop() {
	DEBUG_ENTRY

	struct TTCNetPacketOptOut OptOut;

	memcpy(&OptOut, &m_tOptIn, sizeof(struct TTCNetPacketOptOut));

	Network::Get()->SendTo(m_aHandles[0], &OptOut, sizeof(struct TTCNetPacketOptOut), m_tNode.nIPAddressBroadcast, TCNetBroadcast::PORT_0);

	DEBUG_EXIT
}

void TCNet::HandlePort60000Incoming() {
	const auto packet = &(m_TTCNet.TCNetPacket);
	const auto type  = static_cast<TTCNetMessageType>(packet->ManagementHeader.MessageType);

	DEBUG_PRINTF("MessageType = %d", static_cast<int>(type));

	if (type == TCNET_MESSAGE_TYPE_OPTIN) {
#ifndef NDEBUG
		DumpOptIn();
#endif
		return;
	}
}

void TCNet::HandlePort60001Incoming() {
	if (__builtin_expect((m_pTCNetTimeCode != nullptr), 1)) {
		if (static_cast<TTCNetMessageType>(m_TTCNet.TCNetPacket.ManagementHeader.MessageType) == TCNET_MESSAGE_TYPE_TIME) {
			struct TTCNetTimeCode TimeCode;

			if (!m_bUseTimeCode) {
				uint32_t nTime = *m_pLTime;

				const uint32_t nHours = nTime / 3600000;
				nTime -= nHours * 3600000;
				const uint32_t nMinutes = nTime / 60000;
				nTime -= nMinutes * 60000;
				const uint32_t nSeconds = nTime / 1000;
				const uint32_t nMillis = nTime - nSeconds * 1000;
				const uint32_t nFrames = static_cast<float>(nMillis) / m_fTypeDivider;

				TimeCode.nFrames = nFrames;
				TimeCode.nSeconds = nSeconds;
				TimeCode.nMinutes = nMinutes;
				TimeCode.nHours = nHours;
				TimeCode.nType = m_tTimeCodeType;
			} else {
				TimeCode.nFrames = m_pLTimeCode->Frames;
				TimeCode.nSeconds = m_pLTimeCode->Seconds;
				TimeCode.nMinutes = m_pLTimeCode->Minutes;
				TimeCode.nHours = m_pLTimeCode->Hours;

				uint8_t nSMPTEMode = m_pLTimeCode->SMPTEMode;

				if (nSMPTEMode < 24) {
					nSMPTEMode = m_TTCNet.TCNetPacket.Time.SMPTEMode;
				}

				if (nSMPTEMode == 24) {
					TimeCode.nType = TCNET_TIMECODE_TYPE_FILM;
				} else if (nSMPTEMode == 25) {
					TimeCode.nType = TCNET_TIMECODE_TYPE_EBU_25FPS;
				} else if (nSMPTEMode == 29) {
					TimeCode.nType = TCNET_TIMECODE_TYPE_DF;
				} else if (nSMPTEMode == 30) {
					TimeCode.nType = TCNET_TIMECODE_TYPE_SMPTE_30FPS;
				} else {
					TimeCode.nType = TCNET_TIMECODE_TYPE_EBU_25FPS;
				}
			}

			m_pTCNetTimeCode->Handler(&TimeCode);
		}
	}
}

void TCNet::HandlePort60002Incoming() {
	DEBUG_ENTRY

	DEBUG_EXIT
}

void TCNet::HandlePortUnicastIncoming() {
	DEBUG_ENTRY
#ifndef NDEBUG
	const auto packet = &(m_TTCNet.TCNetPacket);
	const auto type  = static_cast<TTCNetMessageType>(packet->ManagementHeader.MessageType);
#endif
	DEBUG_PRINTF("MessageType = %d", static_cast<int>(type));

	DEBUG_EXIT
}

void TCNet::HandleOptInOutgoing() {
	m_tOptIn.ManagementHeader.SEQ += 1;
	m_tOptIn.ManagementHeader.TimeStamp = Hardware::Get()->Micros();
	m_tOptIn.Uptime = Hardware::Get()->GetUpTime();

	Network::Get()->SendTo(m_aHandles[0], &m_tOptIn, sizeof(struct TTCNetPacketOptIn), m_tNode.nIPAddressBroadcast, TCNetBroadcast::PORT_0);
}

void TCNet::Run() {
	auto packet = reinterpret_cast<uint8_t*>(&m_TTCNet.TCNetPacket);
	uint16_t nForeignPort;

	m_TTCNet.BytesReceived = Network::Get()->RecvFrom(m_aHandles[1], packet, sizeof(m_TTCNet.TCNetPacket), &m_TTCNet.IPAddressFrom, &nForeignPort) ;

	if (m_TTCNet.BytesReceived != 0) {
		HandlePort60001Incoming();
	}

	m_TTCNet.BytesReceived = Network::Get()->RecvFrom(m_aHandles[0], packet, sizeof(m_TTCNet.TCNetPacket), &m_TTCNet.IPAddressFrom, &nForeignPort) ;

	if (m_TTCNet.BytesReceived != 0) {
		HandlePort60000Incoming();
	}

#if defined(USE_PORT_60002)
	m_TTCNet.BytesReceived = Network::Get()->RecvFrom(m_aHandles[2], packet, sizeof(m_TTCNet.TCNetPacket), &m_TTCNet.IPAddressFrom, &nForeignPort) ;

	if (m_TTCNet.BytesReceived != 0) {
		HandlePort60002Incoming();
	}
#endif

#if defined(USE_PORT_UNICAST)
	m_TTCNet.BytesReceived = Network::Get()->RecvFrom(m_aHandles[3], packet, sizeof(m_TTCNet.TCNetPacket), &m_TTCNet.IPAddressFrom, &nForeignPort) ;

	if (m_TTCNet.BytesReceived != 0) {
		HandlePortUnicastIncoming();
	}
#endif

	m_nCurrentMillis = Hardware::Get()->Millis();

	if (__builtin_expect(((m_nCurrentMillis - m_nPreviousMillis) >= 1000), 0)) {
		m_nPreviousMillis = m_nCurrentMillis;
		HandleOptInOutgoing();
	}
}

void TCNet::SetLayer(TCNetLayer tLayer) {
	if(tLayer >= TCNetLayer::LAYER_UNDEFINED) { // TODO Backward compatibility, subject for removal
		tLayer = TCNetLayer::LAYER_M;
		m_bUseTimeCode = true;
	}

	m_tLayer = tLayer;
	m_pLTime = &m_TTCNet.TCNetPacket.Time.L1Time + static_cast<uint32_t>(tLayer);
	m_pLTimeCode =
			reinterpret_cast<struct TTCNetPacketTimeTimeCode*>((reinterpret_cast<uintptr_t>((&m_TTCNet.TCNetPacket.Time.L1TimeCode))
					+ static_cast<uintptr_t>(tLayer) * sizeof(struct TTCNetPacketTimeTimeCode)));
}

void TCNet::SetNodeName(const char *pNodeName) {
#if (__GNUC__ > 8)
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wstringop-truncation"
#endif
	strncpy(reinterpret_cast<char*>(m_tOptIn.ManagementHeader.NodeName), pNodeName, sizeof m_tOptIn.ManagementHeader.NodeName - 1);
	m_tOptIn.ManagementHeader.NodeName[sizeof m_tOptIn.ManagementHeader.NodeName - 1] = '\0';
#if (__GNUC__ > 8)
#pragma GCC diagnostic pop
#endif
}

char TCNet::GetLayerName(TCNetLayer tLayer) {
	switch (tLayer) {
	case TCNetLayer::LAYER_1:
	case TCNetLayer::LAYER_2:
	case TCNetLayer::LAYER_3:
	case TCNetLayer::LAYER_4:
		return static_cast<char>(tLayer) + '1';
		break;
	case TCNetLayer::LAYER_A:
		return 'A';
		break;
	case TCNetLayer::LAYER_B:
		return 'B';
		break;
	case TCNetLayer::LAYER_M:
		return 'M';
		break;
	case TCNetLayer::LAYER_C:
		return 'C';
		break;
	default:
		break;
	}

	return ' ';
}

TCNetLayer TCNet::GetLayer(char nChar) {
	switch (nChar | 0x20) {	// to lower case
	case '1':
	case '2':
	case '3':
	case '4':
		return static_cast<TCNetLayer>(nChar - '1');
		break;
	case 'a':
		return TCNetLayer::LAYER_A;
		break;
	case 'b':
		return TCNetLayer::LAYER_B;
		break;
	case 'm':
		return TCNetLayer::LAYER_M;
		break;
	case 'c':
		return TCNetLayer::LAYER_C;
		break;
	default:
		break;
	}

	return TCNetLayer::LAYER_UNDEFINED;
}

void TCNet::SetTimeCodeType(TTCNetTimeCodeType tType) {
	switch (tType) {
	case TCNET_TIMECODE_TYPE_FILM:
		m_fTypeDivider = 1000.0F / 24;
		break;
	case TCNET_TIMECODE_TYPE_EBU_25FPS:
		m_fTypeDivider = 1000.0F / 25;
		break;
	case TCNET_TIMECODE_TYPE_DF:
		m_fTypeDivider = 1000.0F / 29.97;
		break;
	case TCNET_TIMECODE_TYPE_SMPTE_30FPS:
		m_fTypeDivider = 1000.0F / 30;
		break;
	default:
		return;
		break;
	}

	m_tTimeCodeType = tType;
}

File: RemoteConfigUI/src/org/orangepi/dmx/TCNet.java
/* Copyright (C) 2019-2020 by Arjan van Vught mailto:info@orangepi-dmx.nl
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package org.orangepi.dmx;

import java.awt.Color;
import java.awt.EventQueue;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.SocketException;

import javax.swing.ButtonGroup;
import javax.swing.GroupLayout;
import javax.swing.GroupLayout.Alignment;
import javax.swing.JDialog;
import javax.swing.JRadioButton;
import javax.swing.JTextField;
import javax.swing.LayoutStyle.ComponentPlacement;
import javax.swing.JCheckBox;

public class TCNet extends JDialog {
	private static final long serialVersionUID = -7775636436383674851L;
	private static final int UDP_PORT = 2762;
	
	private InetAddress IPAddressLtcNode; 
    private DatagramSocket socket = null;
    private JTextField txtLayer;
    private JTextField txtType;
    private JRadioButton rdbtnLayer1;
    private JRadioButton rdbtnLayerA;
    private JRadioButton rdbtnLayerB;
    private JRadioButton rdbtnLayerM;
    private JRadioButton rdbtnLayerC;
    private JRadioButton rdbtnLayer4;
    private JRadioButton rdbtnLayer3;
    private JRadioButton rdbtnLayer2;
    private JRadioButton rdbtn30FPS;
    private JRadioButton rdbtn29FPS;
    private JRadioButton rdbtn25FPS;
    private JRadioButton rdbtn24FPS;
    private final ButtonGroup buttonGroupLayer = new ButtonGroup();
    private final ButtonGroup buttonGroupType = new ButtonGroup();
    private JCheckBox chckbxUseTimecode;

	public static void main(String[] args) {
		EventQueue.invokeLater(new Runnable() {
			public void run() {
				try {
					TCNet dialog = new TCNet(InetAddress.getByName("192.168.2.120"));
					dialog.setDefaultCloseOperation(JDialog.DISPOSE_ON_CLOSE);
					dialog.setVisible(true);
				} catch (Exception e) {
					e.printStackTrace();
				}
			}
		});
	}

	public TCNet(InetAddress IPAddressLtcNode) {
		InitComponents();
		CreateEvents();
	
		this.IPAddressLtcNode = IPAddressLtcNode;
		
		 try {
			socket = new DatagramSocket();
			setTitle("TCNet: LTC Node " + IPAddressLtcNode.getHostAddress());
		} catch (SocketException e) {
			setTitle("Error: LTC Node " + IPAddressLtcNode.getHostAddress());
			e.printStackTrace();
		}
	}
	
	public void Show() {
		setVisible(true);
	}
	
	private void InitComponents() {
		setBounds(100, 100, 299, 187);
		
		txtLayer = new JTextField();
		txtLayer.setBorder(null);
		txtLayer.setDragEnabled(false);
		txtLayer.setBackground(new Color(238, 238, 238));
		txtLayer.setText("Layer");
		txtLayer.setEditable(false);
		txtLayer.setColumns(10);
		
		rdbtnLayer1 = new JRadioButton("1");
		buttonGroupLayer.add(rdbtnLayer1);
		
		rdbtnLayer2 = new JRadioButton("2");
		buttonGroupLayer.add(rdbtnLayer2);
		
		rdbtnLayer4 = new JRadioButton("4");
		buttonGroupLayer.add(rdbtnLayer4);
		
		rdbtnLayer3 = new JRadioButton("3");
		buttonGroupLayer.add(rdbtnLayer3);
		
		rdbtnLayerA = new JRadioButton("A");
		buttonGroupLayer.add(rdbtnLayerA);
		
		rdbtnLayerB = new JRadioButton("B");
		buttonGroupLayer.add(rdbtnLayerB);
		
		rdbtnLayerM = new JRadioButton("M");
		buttonGroupLayer.add(rdbtnLayerM);
		
		rdbtnLayerC = new JRadioButton("C");
		buttonGroupLayer.add(rdbtnLayerC);
		
		txtType = new JTextField();
		txtType.setText("Type");
		txtType.setEditable(false);
		txtType.setDragEnabled(false);
		txtType.setColumns(10);
		txtType.setBorder(null);
		txtType.setBackground(new Color(238, 238, 238));
		
		rdbtn24FPS = new JRadioButton("24");
		buttonGroupType.add(rdbtn24FPS);
		
		rdbtn25FPS = new JRadioButton("25");
		buttonGroupType.add(rdbtn25FPS);
		
		rdbtn29FPS = new JRadioButton("29.97");
		buttonGroupType.add(rdbtn29FPS);
		
		rdbtn30FPS = new JRadioButton("30");
		buttonGroupType.add(rdbtn30FPS);
		
		chckbxUseTimecode = new JCheckBox("Use TimeCode");
		
		GroupLayout groupLayout = new GroupLayout(getContentPane());
		groupLayout.setHorizontalGroup(
			groupLayout.createParallelGroup(Alignment.LEADING)
				.addGroup(groupLayout.createSequentialGroup()
					.addGap(18)
					.addGroup(groupLayout.createParallelGroup(Alignment.LEADING)
						.addComponent(chckbxUseTimecode)
						.addGroup(groupLayout.createSequentialGroup()
							.addComponent(txtType, GroupLayout.PREFERRED_SIZE, 48, GroupLayout.PREFERRED_SIZE)
							.addPreferredGap(ComponentPlacement.RELATED)
							.addComponent(rdbtn24FPS)
							.addPreferredGap(ComponentPlacement.RELATED)
							.addComponent(rdbtn25FPS)
							.addGap(2)
							.addComponent(rdbtn29FPS)
							.addPreferredGap(ComponentPlacement.RELATED)
							.addComponent(rdbtn30FPS))
						.addGroup(groupLayout.createSequentialGroup()
							.addComponent(txtLayer, GroupLayout.PREFERRED_SIZE, 48, GroupLayout.PREFERRED_SIZE)
							.addPreferredGap(ComponentPlacement.RELATED)
							.addGroup(groupLayout.createParallelGroup(Alignment.LEADING)
								.addComponent(rdbtnLayer1)
								.addComponent(rdbtnLayerA, GroupLayout.PREFERRED_SIZE, 46, GroupLayout.PREFERRED_SIZE))
							.addPreferredGap(ComponentPlacement.RELATED)
							.addGroup(groupLayout.createParallelGroup(Alignment.TRAILING)
								.addComponent(rdbtnLayer2, GroupLayout.PREFERRED_SIZE, 40, GroupLayout.PREFERRED_SIZE)
								.addComponent(rdbtnLayerB, GroupLayout.PREFERRED_SIZE, 40, GroupLayout.PREFERRED_SIZE))
							.addPreferredGap(ComponentPlacement.RELATED)
							.addGroup(groupLayout.createParallelGroup(Alignment.LEADING)
								.addComponent(rdbtnLayer3, GroupLayout.PREFERRED_SIZE, 40, GroupLayout.PREFERRED_SIZE)
								.addComponent(rdbtnLayerM))
							.addGap(9)
							.addGroup(groupLayout.createParallelGroup(Alignment.LEADING)
								.addComponent(rdbtnLayerC)
								.addComponent(rdbtnLayer4, GroupLayout.PREFERRED_SIZE, 40, GroupLayout.PREFERRED_SIZE))))
					.addGap(68))
		);
		groupLayout.setVerticalGroup(
			groupLayout.createParallelGroup(Alignment.LEADING)
				.addGroup(groupLayout.createSequentialGroup()
					.addGroup(groupLayout.createParallelGroup(Alignment.LEADING)
						.addGroup(groupLayout.createSequentialGroup()
							.addGap(16)
							.addGroup(groupLayout.createParallelGroup(Alignment.LEADING)
								.addComponent(rdbtnLayer1)
								.addGroup(groupLayout.createParallelGroup(Alignment.BASELINE)
									.addComponent(rdbtnLayer2)
									.addComponent(rdbtnLayer3)
									.addComponent(rdbtnLayer4)))
							.addPreferredGap(ComponentPlacement.UNRELATED)
							.addGroup(groupLayout.createParallelGroup(Alignment.LEADING)
								.addComponent(rdbtnLayerC)
								.addComponent(rdbtnLayerM)
								.addComponent(rdbtnLayerB)
								.addComponent(rdbtnLayerA)))
						.addGroup(groupLayout.createSequentialGroup()
							.addGap(34)
							.addComponent(txtLayer, GroupLayout.PREFERRED_SIZE, GroupLayout.DEFAULT_SIZE, GroupLayout.PREFERRED_SIZE)))
					.addPreferredGap(ComponentPlacement.UNRELATED)
					.addGroup(groupLayout.createParallelGroup(Alignment.LEADING)
						.addComponent(txtType, GroupLayout.PREFERRED_SIZE, 16, GroupLayout.PREFERRED_SIZE)
						.addGroup(groupLayout.createParallelGroup(Alignment.BASELINE)
							.addComponent(rdbtn24FPS)
							.addComponent(rdbtn25FPS)
							.addComponent(rdbtn29FPS)
							.addComponent(rdbtn30FPS)))
					.addGap(18)
					.addComponent(chckbxUseTimecode)
					.addContainerGap(62, Short.MAX_VALUE))
		);
		getContentPane().setLayout(groupLayout);
	}
	
	private void CreateEvents() {
		rdbtnLayer1.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				SendUpd("tcnet!layer#1");
			}
		});
		
		rdbtnLayer2.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				SendUpd("tcnet!layer#2");
			}
		});
		
		rdbtnLayer3.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				SendUpd("tcnet!layer#3");
			}
		});
		
		rdbtnLayer4.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				SendUpd("tcnet!layer#4");
			}
		});
		
		rdbtnLayerA.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				SendUpd("tcnet!layer#A");
			}
		});
		
		rdbtnLayerB.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				SendUpd("tcnet!layer#B");
			}
		});
		
		rdbtnLayerM.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				SendUpd("tcnet!layer#M");
			}
		});
		
		rdbtnLayerC.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				SendUpd("tcnet!layer#C");
			}
		});
		
		rdbtn24FPS.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				SendUpd("tcnet!type#24");
			}
		});
		
		rdbtn25FPS.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				SendUpd("tcnet!type#25");
			}
		});
		
		rdbtn29FPS.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				SendUpd("tcnet!type#29");
			}
		});
		
		rdbtn30FPS.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				SendUpd("tcnet!type#30");
			}
		});
		
		chckbxUseTimecode.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent arg0) {
				if (chckbxUseTimecode.isSelected()) {
					SendUpd("tcnet!timecode#y");
					DisablerdbtnFPS();
				} else {
					SendUpd("tcnet!timecode#n");
					EnablerdbtnFPS();
				}
			}
		});
	}
	
	private void DisablerdbtnFPS() {
		rdbtn24FPS.setEnabled(false);
		rdbtn25FPS.setEnabled(false);
		rdbtn29FPS.setEnabled(false);
		rdbtn30FPS.setEnabled(false);
	}

	private void EnablerdbtnFPS() {
		rdbtn24FPS.setEnabled(true);
		rdbtn25FPS.setEnabled(true);
		rdbtn29FPS.setEnabled(true);
		rdbtn30FPS.setEnabled(true);
	}

	private void SendUpd(String message) {
		byte[] buffer = message.getBytes();
		
		DatagramPacket UDPPacket = new DatagramPacket(buffer, buffer.length, IPAddressLtcNode, UDP_PORT);
		try {
			socket.send(UDPPacket);
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
}
